chainerrl/envs/multiprocess_vector_env.py

```diff
@@ -0,0 +1,134 @@
+        env_fns (list of callable): List of callables, each of which
+            returns gym.Env that is run in its own subprocess.
+    """
+
+    def __init__(self, env_fns):
+        nenvs = len(env_fns)
+        self.remotes, self.work_remotes = zip(*[Pipe() for _ in range(nenvs)])
+        self.ps = \
+            [Process(target=worker, args=(work_remote, env_fn))
+             for (work_remote, env_fn) in zip(self.work_remotes, env_fns)]
+        for p in self.ps:
+            p.start()
+        self.last_obs = [None] * self.num_envs
+        self.remotes[0].send(('get_spaces', None))
+        self.action_space, self.observation_space = self.remotes[0].recv()
+        self.closed = False
+
+    def __del__(self):
+        if not self.closed:
+            self.close()
+
+    @cached_property
+    def spec(self):
+        self._assert_not_closed()
+        self.remotes[0].send(('spec', None))
+        spec = self.remotes[0].recv()
+        return spec
+
+    def step(self, actions):
+        self._assert_not_closed()
+        for remote, action in zip(self.remotes, actions):
+            remote.send(('step', action))
+        results = [remote.recv() for remote in self.remotes]
+        self.last_obs, rews, dones, infos = zip(*results)
+        return self.last_obs, rews, dones, infos
+
+    def reset(self, mask=None):
+        self._assert_not_closed()
+        if mask is None:
+            mask = np.zeros(self.num_envs)
+        for m, remote in zip(mask, self.remotes):
+            if not m:
+                remote.send(('reset', None))
+
+        obs = [remote.recv() if not m else o for m, remote,
+               o in zip(mask, self.remotes, self.last_obs)]
+        self.last_obs = obs
+        return obs
+
+    def close(self):
+        self._assert_not_closed()
+        self.closed = True
+        for remote in self.remotes:
+            remote.send(('close', None))
+        for p in self.ps:
+            p.join()
+
+    def seed(self, seeds=None):
+        self._assert_not_closed()
+        if seeds is not None:
+            if isinstance(seeds, int):
+                seeds = [seeds] * self.num_envs
+            elif isinstance(seeds, list):
+                if len(seeds) != self.num_envs:
+                    raise ValueError(
+                        "length of seeds must be same as num_envs {}"
+                        .format(self.num_envs))
+            else:
+                raise TypeError(
+                    "Type of Seeds {} is not supported.".format(type(seeds)))
+        else:
+            seeds = [None] * self.num_envs
+
+        for remote, seed in zip(self.remotes, seeds):
+            remote.send(('seed', seed))
+        results = [remote.recv() for remote in self.remotes]
+        return results
+
+    @property
+    def num_envs(self):
+        return len(self.remotes)
+
+    def _assert_not_closed(self):
+        assert not self.closed, "Trying to operate on a MultiprocessVectorEnv after calling close()"  # NOQA
```

Comment:
The message might be confusing when `close` is called twice.

Reply:
simplified the message.